File: cpp/Week-1-Number-Guessing/tests/test_number_guess.cpp
#include <cassert>
#include <vector>
#include <iostream>
using namespace std;

string checkGuess(int secret, int guess) {
    // Paste the same implementation of checkGuess from number_guess.cpp here
}

int calculateScore(int attempts) {
    // Paste the same implementation of calculateScore from number_guess.cpp here

}

string giveHint(int secret, vector<int> history) {
    // Paste the same implementation of giveHint from number_guess.cpp here
}

int main() {

    assert(checkGuess(50, 70) == "HIGH");
    assert(checkGuess(50, 30) == "LOW");
    assert(checkGuess(50, 50) == "CORRECT");

    assert(calculateScore(2) == 100);
    assert(calculateScore(5) == 70);
    assert(calculateScore(8) == 40);
    assert(calculateScore(12) == 10);

    vector<int> h1 = {20, 80};
    assert(giveHint(60, h1) == "Try between 21 and 79");

    vector<int> h2 = {10, 90, 40, 70};
    assert(giveHint(60, h2) == "Close!");

    vector<int> h3 = {5, 95, 10};
    assert(giveHint(50, h3) == "Try between 11 and 94");

    vector<int> h4 = {10, 90, 55};
    assert(giveHint(60, h4) == "Very Close!");

    cout << "All tests passed!" << endl;

    return 0;
}

File: python/Week-1-Number-Guessing/tests/test_number_guess.py
import unittest
from number_guess import check_guess, calculate_score, give_hint


class TestNumberGuess(unittest.TestCase):

    def test_check_guess(self):
        self.assertEqual(check_guess(50, 70), "HIGH")
        self.assertEqual(check_guess(50, 30), "LOW")
        self.assertEqual(check_guess(50, 50), "CORRECT")

    def test_calculate_score(self):
        self.assertEqual(calculate_score(2), 100)
        self.assertEqual(calculate_score(5), 70)
        self.assertEqual(calculate_score(8), 40)
        self.assertEqual(calculate_score(15), 10)

    def test_basic_hints(self):
        self.assertEqual(give_hint(60, [58]), "Very Close!")
        self.assertEqual(give_hint(60, [52]), "Close!")

    def test_history_range_narrowing(self):
        history = [20, 80]
        self.assertEqual(
            give_hint(60, history),
            "Try between 21 and 79"
        )

    def test_multiple_constraints(self):
        history = [10, 90, 40, 70]
        self.assertEqual(
            give_hint(60, history),
            "Close!"
        )

    def test_last_guess_only_is_wrong(self):
        history = [5, 95, 10]
        self.assertEqual(
            give_hint(50, history),
            "Try between 11 and 94"
        )

    def test_close_priority(self):
        history = [10, 90, 55]
        self.assertEqual(
            give_hint(60, history),
            "Very Close!"
        )


if __name__ == "__main__":
    unittest.main()
